### ISSUE TITLE
[Feature Request] Run on background process. / [Bug] UI freeze when run.

### ISSUE THREAD

**Issue by lkhphuc** [denolehov/obsidian-git#69]
I set git backup to run automatically every hour.
Everytime it runs, Obsidian becomes laggy and unresponsive until it finish, which can takes up to 30s-60s. If I happen to be using Obsidian at this exact time 2-3 times in a morning, it can be very frustrating.
I wonder if this process can be run in the background? 

Maybe related if #68 is fixed.

**Comment by MrVPlusOne** [comment]
I'm having the same issue here! One thing that feels strange to me is that if I manually push the changes from within the command line using `git`, it takes almost no time for the push to finish, whereas if I run the push command of the plugin from inside Obsidian, it freezes the UI and can take more than 30 seconds. What extra work does the plugin do that makes it so much slower than `git`?

**Comment by 23tux** [comment]
Same issue here! I've configured obsidian-git to push changes every 5 minutes, so the lagging / freezing becomes REALLY annoying.

**Comment by broven** [comment]
Same issue

**Comment by dsteinbock** [comment]
Same issue. When Obsidian Git is running, the Obsidian process spikes to 90% CPU for ~30 seconds.

When What OS are y'all running? I didn't notice this until after I upgraded from MacOS Catalina to Big Sur last week.

**Comment by MrVPlusOne** [comment]
Actually, I think this issue was greatly improved after I upgraded my Obsidian and the plugin to the newest version. Maybe you can try to upgrade your Obsidian under the "Options -> About" tab.

**Comment by akselw** [comment]
I have the same issue, and I experience it on both my Mac running Catalina and my Mac running Big Sur. I just noticed that Ben Hong had the same issue on his live stream a couple of days ago. You can see it, and he points it out at the 32:30 mark here: https://www.twitch.tv/videos/1044660956?t=0h32m30s

**Comment by lkhphuc** [comment]
I updated the issue's name to reflect two side of this issue.

**Comment by gabros20** [comment]
I have the same issue on Big Sur. Freezes the UI until the commit runs for 10-15 seconds. Recently I've updated Mac OS High Sierra to Big Sure and thats when I noticed the issue straight after.

**Comment by dwSun** [comment]
Same issue,
Big Sur 11.4
Obsidian v0.12.5
Obsidian Git 1.9.2

I have 1625 files, most are pdfs, all those files take 1.7GB on my disk.

UI freezes for 10~15 seconds when starting up or pushing to github.

![image](https://user-images.githubusercontent.com/5899962/124051237-ae35d580-da4e-11eb-8106-3e520493e716.png)

**Comment by MrModest** [comment]
same issue. and it's only on mac. on Windows 10 works fine, but on Mac freezing even I open setting page for plugin

**Comment by prakar** [comment]
Same issue. When the plugin is enabled, CPU maxes out and Obsidian can't be used. Every click takes a long time to process.

On OSX 10.11.8 El Capitan

![Screen Shot 2021-07-23 at 6 09 27 PM](https://user-images.githubusercontent.com/811452/126782880-f26ed285-9674-42bd-a0f0-bcab1383cc5b.png)
![Screen Shot 2021-07-23 at 6 08 24 PM](https://user-images.githubusercontent.com/811452/126782895-1f343d91-f555-46e1-b869-ffec5ab0afcc.png)

**Comment by ronilaukkarinen** [comment]
Yeah this happens for me as well on macOS (Big Sur). On Windows 10 I have no issues.

**Comment by Vinzent03** [comment]
I am only on Linux and Windows and I think you are all macOS users right? It's hard to reproduce for me. I think the problem must be somewhere else. I don't know if running it in a different process is the right way and would even fix it.

**Comment by MrModest** [comment]
@Vinzent03 can I help you get some logs for diagnose problem?

**Comment by MrModest** [comment]
>  think the problem must be somewhere else.
but it reproduce only for this plugin.

**Comment by MrModest** [comment]
https://user-images.githubusercontent.com/13031058/130359938-d15a6aa7-b4ca-48ea-b4b8-81914811a2e1.mov

**Comment by MrModest** [comment]
What common have launch settings page and commit? May be it relate with git library? On mac git require Xcode
May be some problem relate with that.

**Comment by Vinzent03** [comment]
I think the root of the issue is either the communication of `simple-git`(external library to communicate with local git installation) with git or a wrong/bad git installation. I am not sure what further details you could provide me.

**Comment by MrModest** [comment]
@Vinzent03 may be you can add some verbose logs for localize the problem, and I will run this and give this logs for you.
For understanding, freeze happens inside lib method or your code. May it some concrete freeze method and you can find alternative.

Your plugin is very comfortable. It would be great continue use it on Mac

**Comment by Vinzent03** [comment]
Sorry for late response. Seems like every communication with simpe-git/local git installation freezes Obsidian. I don't think I will have the time to put that in another process (if it works at all). Pull requests are welcome 😅

**Comment by olearycrew** [comment]
I would think the only way around this is to spawn another process - I wonder if that's even possible with the Obsidian plugin infrastructure.

**Comment by MrModest** [comment]
some funny moment. In Macbook Pro with M1 it still freezing, but now it about 1 second or a bit smaller instead 2 or more seconds in Intel version :D

**Comment by dsteinbock** [comment]
Not sure it's been said already in this thread: the plugin did not have this behavior in the past, on MacOS. So either a change to Obsidian or a change to the plugin introduced this issue. Has that breaking change been tracked down?

**Comment by warrenalphonso** [comment]
@dsteinbock Do you remember roughly how long ago it worked fine? Maybe try a past release around then and see if it persists? https://github.com/denolehov/obsidian-git/tags

**Comment by KierenB** [comment]
I've also been having this issue (Intel Mac, plugin version 1.11), for me the pause is 6-7sec so really uncomfortable in terms of blocking note-taking at important times. For the time being I've gone back to the background task of uploading to GitHub as described here: https://medium.com/analytics-vidhya/how-i-put-my-mind-under-version-control-24caea37b8a5

kudos to @Vinzent03 for an excellent plugin, hopefully someone with more skill than me can help him with a solution 😅

**Comment by davecan** [comment]
Confirming same issue here. Freeze is about 40-50 seconds so I've had to set the sync to 180 minutes just to make it tolerable. Preferably it would sync much sooner than that.

macOS Big Sur 11.6, Obsidian v0.12.15, git version 2.32.

Interesting to note that I currently have 3 instances of Obsidian open, each pointing to a separate vault, and they have been open for multiple days. I also have it set to pull before push and to show the status bar updates.

I closed all of them, forced a restart of Obsidian, and tried changing a couple of settings.

- Turned off pull before push, lag is reduced only a bit. And even though the status bar says the operation is complete and the UI responds, the UI is still extremely slow in opening the settings once a push occurs.

- Then turned off the status bar updates and it is still lagging.

After the second change, measuring the time from initiation of the commit to the time when the note screen is responsive again is 32 seconds. But when I immediately open the settings screen right after the popup message says the git operation is complete the UI still lags until nearly 45 seconds (total) before the settings UI is usable.

Interestingly, the status bar says it is now 4 minutes since the last git update, and when I just reopened the settings screen it once again lagged for about 10 seconds before even displaying the screen. The Obsidian Helper (Renderer) process spiked the CPU as described by others. 

I did have the git plugin configured to run every 1 minute for the tests I described above so it appears there is some interaction happening here that is causing effects that persist beyond the execution of the plugin itself.

When I switched the plugin back to 180 minute intervals and then terminated the entire Obsidian process and restarted it with a single vault, the expected smooth fluid operations returned, no noticeable lag in usage.

**Comment by dsteinbock** [comment]
> @dsteinbock Do you remember roughly how long ago it worked fine? Maybe try a past release around then and see if it persists?

Well the issue was first reported by @lkhphuc on April 21. I remember dealing with the lag for weeks — thinking it was an Obsidian bug that would get fixed — until I traced it to this plugin.

So we should try testing old versions of Obsidian w/ old versions of obsidian-git to backtrace. 

e.g. 
Obsidian release from March 21: https://github.com/obsidianmd/obsidian-releases/releases/tag/v0.11.9
obsidian-git from March 17: https://github.com/denolehov/obsidian-git/releases/tag/1.6.1

other Obsidian releases:
https://github.com/obsidianmd/obsidian-releases/releases?after=v0.11.11

other obsidian-git releases:
https://github.com/denolehov/obsidian-git/releases?after=1.9.0

**Comment by miluoshi** [comment]
This happens to me as well.

I've tried debugging this issue with Chrome DevTools. 

While the backup task is running there are multiple `ChildProcess.spawn` commands being run by `simple-git` library, each spawn taking ~3s to finish and each blocking UI thread.
![image](https://user-images.githubusercontent.com/1130547/135314419-00beb07a-12ac-4d1d-9b36-fce623df05ce.png)

Googling directed me to [issue in electron repo](https://github.com/electron/electron/issues/26143), that seems to be related to this issue.

### My system:
- **OS:** MacOS Big Sur 11.6
- **Obsidian:** v0.12.15 (Electron v12.0.18)
- **Obsidian Git:** v1.13.0

**Comment by 23tux** [comment]
Is there any news on this? I've increased the committing interval to 1h, but it would be nice to have a more fine grained history about my edits

**Comment by kmf** [comment]
Not sure if it's related, I have/had the same issue,
I have `Homebrew` installed, and was using the updated git that shipped with it
```
git 2.34.1
```

I did a simple 
```
brew unlink git
Unlinking /usr/local/Cellar/git/2.34.1... 213 symlinks removed.
```

When I then ran
```
git --version
git version 2.32.0 (Apple Git-132)
```

Related Link: https://www.jeffgeerling.com/comment/17150#comment-17150

### My system:
- OS: MacOS Monterey 12.2
- Obsidian: v0.13.19
- Obsidian Git: v1.20.0

**Comment by aoxuely** [comment]
I have the same issue，but I founded that while obsidian doing backup works,the cpu also cost much
[image](https://user-images.githubusercontent.com/10103217/150670099-45dc775a-2e2e-4aa5-a0e1-cc7bf75a1426.png)

**Comment by RumiAllbert** [comment]
My solution to this was to simply write a contab to run git every x number of minutes. I love the plugin but will have to wait until there is a fix.

**Comment by miluoshi** [comment]
Today I was prototyping an idea that could partially solve this freezing issue and it seems it might work.

I was thinking it would be great if the plugin did auto backups only when Obsidian window is in background (not focused).
It would be much less annoying if auto backup were the freezes happening only when I am not currently using Obsidian.

It can be implemented using `blur` and `focus` events on window. For tabbing between fullscreen apps also `visibilitychange` event works.
Auto backups would run only X minutes after `blur` and on `focus` the timer would be cancelled. What do you think?

Pseudo code:
```js
const backgroundBackupTimeout = 1000 * 60 * X; // X minutes
let timerId;

window.addEventListener('blur', () => {
  timerId = setTimeout(() => {
    // do backup
  }, backgroundBackupTimeout);
});

window.addEventListener('focus', () => {
  if (timerId) {
    clearTimeout(timerId);
    timerId = undefined;
  }
});
```

**Comment by davecan** [comment]
Speaking only for myself I think this could be a great thing to try. I typically leave obsidian open most of the time and it is idle in the background most of that time so that would be a good UX experience for me. 

I'm assuming the option to display a notification when the plugin starts would still work, so if a user unblurs the app while it is executing they would see the notice explaining why it is temporarily sluggish?

**Comment by wesmoncrief** [comment]
I would also really appreciate the background solution.

In particular, I often open Obsidian in the middle of a meeting to quickly jot something down, and having it hang for the first ~5+ seconds is pretty annoying.

@Vinzent03 would you be open to a PR that does something along the lines of what @miluoshi suggested?

**Comment by miluoshi** [comment]
I've created a PR with suggested feature above https://github.com/denolehov/obsidian-git/pull/188

**Comment by tht** [comment]
@miluoshi: Thanks a lot… I've just installed it manually in my Obsidian and will report back if there are any issues.

**Comment by miluoshi** [comment]
> @miluoshi: Thanks a lot… I've just installed it manually in my Obsidian and will report back if there are any issues. 

Thanks you! It definitely needs some testing, because I might have missed some edge cases.

**Comment by tht** [comment]
For me it is not working at all. It completely stopped committing and pushing (haven't checked pulling). Window was inactive for hours today but it did not push the local changes.
![image](https://user-images.githubusercontent.com/2557225/155859309-e1f72001-b2b2-469f-962a-f9295f888c12.png)

**Comment by miluoshi** [comment]
> For me it is not working at all. It completely stopped committing and pushing (haven't checked pulling). Window was inactive for hours today but it did not push the local changes.

@tht thanks for testing. I'll get back to it this weekend.

**Comment by miluoshi** [comment]
Notice that fix for the original bug with freezing in libuv was merged yesterday and we're waiting for Electron to merge this fix and release updated version: https://github.com/electron/electron/issues/26143#issuecomment-1057583242

**Comment by samuelskeirik** [comment]
After Electron merges this fix, will we have to wait on Obsidian to also update? Or will that fix in Electron propagate down to Obsidian?

**Comment by fpigeonjr** [comment]
same issue, watching to try again when this is resolved.

**Comment by depressiveRobot** [comment]
After some testing with v0.13.33 (installer v0.13.31), it seems that the problem with the freezing UI is gone. At least for me. Maybe some others can also test and confirm it.

**Comment by miluoshi** [comment]
I am still encountering this issue in v0.14.5 (installer v0.13.24). 

Meanwhile the fix was released in [Electron 16.2.0](https://github.com/electron/electron/issues/26143#issuecomment-1081485875) (current Obsidian v0.14.5 uses Electron 16.0.8) so I've [asked](https://discord.com/channels/686053708261228577/716028884885307432/960171861210193951) in Obsidian Discord if they plan to update Electron soon to fix this bug. Hopefully they update it soon.

**Comment by miluoshi** [comment]
The issue is fixed with Obsidian installer v0.14.5 🎉

**Comment by Vinzent03** [comment]
I hope this issue is now resolved.

**Comment by mogigi** [comment]
I'm on mac Monterey  12.5, using obsidian desktop install v0.15.9, and the obsidian-git plugin installed from community plugins causes this issue of freezing obsidian UI today, just like the past comments from above. The activity monitor, is showing Obsidian Helper (renderer) CPU at 99.8+ percent.
